File: engine_py/tests/test_save_load.py
def test_save_and_load(make_world, tmp_path):
    world = make_world()
    world.set_mode("roguelike")
    e = world.spawn_entity()
    world.set_component(e, "Health", {"current": 99, "max": 100})

    save_file = tmp_path / "test_save.json"
    world.save_to_file(str(save_file))

    world.despawn_entity(e)
    assert len(world.get_entities()) == 0

    world.load_from_file(str(save_file))
    entities = world.get_entities()
    assert len(entities) > 0
    h = world.get_component(entities[0], "Health")
    assert h["current"] == 99

File: engine/scripts/lua/tests/test_save_load.lua
local assert = require("assert")
local utils = require("utils")

local function test_save_and_load()
	-- Setup: create some entities and save them
	local e1 = spawn_entity()
	set_inventory(e1, { slots = utils.empty_array(), weight = 0.0, volume = 0.0 })
	local e2 = spawn_entity()
	set_component(e2, "Health", { current = 10, max = 10 })

	save_to_file("test_save.json")
	-- Despawn all entities in the world
	local entities = get_entities()
	for _, eid in ipairs(entities) do
		despawn_entity(eid)
	end

	local entities_after_despawn = get_entities()
	assert.equals(#entities_after_despawn, 0, "Entities should be empty after despawn")

	-- Restore
	load_from_file("test_save.json")
	local entities_after = get_entities()
	assert.is_true(#entities_after >= 2, "Entities should exist after loading")
end

return {
	test_save_and_load = test_save_and_load,
}
